File: src/java/magma/app/io/source/PathSource.java
package magma.app.io.source;

import magma.api.io.PathLike;
import magma.api.result.Result;
import magma.app.io.location.Location;
import magma.app.io.location.SimpleLocation;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public record PathSource(PathLike sourceDirectory, PathLike source) implements Source {
    private static List<String> computeNamespace(PathLike parent) {
        final List<String> segments = new ArrayList<>();
        for (var i = 0; i < parent.getNameCount(); i++)
            segments.add(parent.getName(i)
                    .asString());
        return segments;
    }

    @Override
    public Result<String, IOException> readString() {
        return this.source.readString();
    }

    @Override
    public Location computeLocation() {
        final var relative = this.sourceDirectory()
                .relativize(this.source);
        final var relativeParent = relative.getParent();

        final var segments = computeNamespace(relativeParent);
        final var namespace = String.join(".", segments);

        final var fileName = this.source.getFileName()
                .asString();
        final var separator = fileName.lastIndexOf(".");
        final var name = fileName.substring(0, separator);

        return new SimpleLocation(namespace, name);
    }
}

File: src/clang/magma/compile/source/PathSource.c
#include "PathSource.h"
struct List__String computeNamespace(){Path_ relative = sourceDirectory.relativize(source);
        Path_ parent = relative.getParent();return parent.stream().collect(ListCollector_());
}
struct String computeName(){String nameWithExt = source().getFileName().asString();return nameWithExt.substring(0, nameWithExt.lastIndexOf("."));
}
struct Result_String_IOError readString(){return source.readString();
}

